README.md
# Linglong Loader

Linglong Loader

## Getting help

Any usage issues can ask for help via

- [Gitter](https://gitter.im/orgs/linuxdeepin/rooms)
- [IRC channel](https://webchat.freenode.net/?channels=deepin)
- [Forum](https://bbs.deepin.org)
- [WiKi](https://wiki.deepin.org/)

## Getting involved

We encourage you to report issues and contribute changes

- [Contribution guide for developers](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers-en)
  . (English)
- [开发者代码贡献指南](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers) (中文)

## License

This project is licensed under [LGPL-3.0-or-later](LICENSE).


debian/changelog
linglong-loader (1.3.3.10-1) unstable; urgency=medium

  * update version for deepin.

 -- huqinghong <huqinghong@uniontech.com>  Tue, 27 Dec 2022 14:37:59 +0800

linglong-loader (1.3.3.9-1) unstable; urgency=medium

  * update version for deepin.

 -- yuanqiliang <yuanqiliang@uniontech.com>  Thu, 25 Nov 2022 14:51:55 +0800

linglong-loader (1.3.3.8-1) unstable; urgency=medium

  * update version for deepin.

 -- huqinghong <huqinghong@uniontech.com>  Thu, 24 Nov 2022 14:51:55 +0800

linglong-loader (1.3.3.7-1) unstable; urgency=medium

  * update version for deepin.

 -- Chen Linxuan <chenlinxuan@uniontech.com>  Fri, 4 Nov 2022 15:00:00 +0800

linglong-loader (1.3.3.6-1) unstable; urgency=medium

  * update version for deepin.

 -- Chen Linxuan <chenlinxuan@uniontech.com>  Thu, 27 Aug 2022 15:00:00 +0800

linglong-loader (1.3.3.5-1) unstable; urgency=medium

  * update version for deepin.

 -- liujianqiang <liujianqiang@uniontech.com>  Mon, 01 Aug 2022 20:56:44 +0800

linglong-loader (1.3.2-1) unstable; urgency=medium

  * update version.

 -- liujianqiang <liujianqiang@uniontech.com>  Fri, 24 Jun 2022 17:01:58 +0800

linglong-loader (1.1.0-1) unstable; urgency=medium

  * update version.

 -- liujianqiang <liujianqiang@uniontech.com>  Fri, 15 Apr 2022 18:50:58 +0800

linglong-loader (1.0-1) unstable; urgency=medium

  * Init package.

 -- chenhuixing <chenhuixing@uniontech.com>  Thu, 09 Sep 2021 11:41:48 +0000


tests/demo_test.cpp
/*
 * SPDX-FileCopyrightText: 2022 UnionTech Software Technology Co., Ltd.
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

#include <iostream>

#include <gtest/gtest.h>

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QProcess>

namespace linglong::squashfs {
int MountFile(const std::string &path, unsigned long offset, const std::string &mount_point);
}

namespace linglong::elf {
long GetELFSize(const std::string &path);
}

TEST(DemoTest, demo01)
{
    QFileInfo fs("../../tests/data/org.deepin.demo_1.2.1_x86_64.uab");
    EXPECT_EQ(fs.exists(), true);
    auto size = linglong::elf::GetELFSize("../../tests/data/org.deepin.demo_1.2.1_x86_64.uab");
    std::cout << size << std::endl;
    EXPECT_EQ(size, 2326904);
    std::string mountPoint("/tmp/uab_test");
    QDir("/tmp/uab_test").mkpath(".");
    //fuse mount uab
    auto result = linglong::squashfs::MountFile("../../tests/data/org.deepin.demo_1.2.1_x86_64.uab", size, mountPoint);
    //fuse umount uab
    QProcess process;
    process.setProgram("fusermount3");
    process.setStandardOutputFile("/dev/null");
    process.setStandardErrorFile("/dev/null");
    process.setArguments({"-u",QString("/tmp/uab_test")});
    process.startDetached();
    EXPECT_EQ(result,0);
}


src/main.cpp
/*
 * SPDX-FileCopyrightText: 2022 UnionTech Software Technology Co., Ltd.
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

#include <pthread.h>
#include <sys/prctl.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <limits.h>
#include <unistd.h>

#include <cstdlib>
#include <iostream>

namespace linglong {

namespace elf {

off_t get_bundle_offset(const std::string &path);

}

namespace erofs {

int MountFile(const std::string &path, size_t offset, const std::string &mount_point);

} // namespace erofs

} // namespace linglong

std::string realpath(const std::string &path)
{
    char *buffer = new char[PATH_MAX];
    auto ret = ::realpath(path.c_str(), buffer);
    if (ret != buffer) {
        std::cerr << "realpath failed:" << errno;
    }
    std::string res = ret;
    delete[] buffer;
    return res;
}

bool exists(const std::string &path)
{
    return access(path.c_str(), F_OK) == 0;
}

int main(int argc, char **argv)
{
    prctl(PR_SET_PDEATHSIG, SIGKILL);

    auto exec_path = realpath("/proc/self/exe");

    if (!exists(exec_path)) {
        std::cerr << "load failed! " << exec_path << std::endl;
        return EXIT_FAILURE;
    }

    // get elf size
    auto bundle_offset = linglong::elf::get_bundle_offset(exec_path);

    if (bundle_offset == -1) {
        std::cerr << "failed to find bundle" << std::endl;
        return EXIT_FAILURE;
    }

    // FIXME: user runtime dir
    char temp_prefix[1024] = "/tmp/uab-XXXXXX";
    char *mount_point = mkdtemp(temp_prefix);

    linglong::erofs::MountFile(exec_path, bundle_offset, mount_point);

    std::string loader_path = mount_point;
    loader_path += "/loader";
    if (!exists(loader_path)) {
        std::cerr << "can not found" << loader_path << std::endl;
        return EXIT_FAILURE;
    }

    auto pid = fork();
    if (pid < 0) {
        std::cerr << "fork exec failed:" << errno << std::endl;
        return EXIT_FAILURE;
    }

    // run loader for child
    if (pid == 0) {
        if (chdir(mount_point) != 0) {
            std::cerr << "change to work failed !" << errno << std::endl;
            return EXIT_FAILURE;
        }

        int cnt = 1;
        // FIXME, user ARG_MAX
        char *load_argv[16] = {
            (char *)loader_path.data(),
            mount_point,
        };
        if (argc > 1 && argc < 16) {
            for (auto idx = 1; idx < argc; idx++) {
                cnt++;
                load_argv[cnt] = argv[idx];
            }
        }
        cnt++;
        load_argv[cnt] = nullptr;
        extern char **environ;
        // run loader shell
        auto ret = execve(loader_path.data(), load_argv, environ);
        if (ret != 0) {
            std::cerr << "run failed!" << std::endl;
        }
        exit(0);
    }

    while (true) {
        int childPid = waitpid(-1, nullptr, 0);
        if (childPid > 0) {
            // std::cout << "wait " << childPid << " exit" << endl;
        }
        if (childPid < 0) {
            std::cout << "all child exit" << std::endl;
            return EXIT_SUCCESS;
        }
    }
}


src/erofs.cpp
/*
 * SPDX-FileCopyrightText: 2022 UnionTech Software Technology Co., Ltd.
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

#include <sys/sem.h>
#include <sys/wait.h>
#include <unistd.h>

#include <array>
#include <iostream>
#include <memory>
#include <string>

extern "C" {
extern int erofs_fuse_main(int argc, char *argv[]);
}

namespace linglong::erofs {

int MountFile(const std::string &path, unsigned long offset, const std::string &mount_point)
{
    int fusefs_pid = vfork();

    if (0 == fusefs_pid) {
        // child
        char *dir = realpath(path.c_str(), nullptr);

        char options[128];
        sprintf(options, "--offset=%lu", offset);

        const char *fuse_argv[] = {dir, options, dir, mount_point.c_str(), nullptr};

        if (0 != erofs_fuse_main(4, const_cast<char **>(fuse_argv))) {
            std::cerr << "erofs_fuse_main" << std::endl;
            return -1;
        }
    }

    waitpid(fusefs_pid, nullptr, 0);

    return 0;
}

} // namespace linglong::erofs


src/squashfs.cpp
/*
 * Copyright (c) 2021. Uniontech Software Ltd. All rights reserved.
 *
 * Author:     Iceyer <me@iceyer.net>
 *
 * Maintainer: Iceyer <me@iceyer.net>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include <unistd.h>
#include <sys/sem.h>

#include <string>
#include <array>
#include <memory>
#include <iostream>

extern "C" {
extern int fusefs_main(int argc, char *argv[], void (*mounted)());
}

namespace linglong::squashfs {

key_t sem_key;

union semun {
    int val;
    struct semid_ds *buf;
    ushort *array;
};

struct sembuf sem_lock = {0, -1, SEM_UNDO};

struct sembuf sem_unlock = {0, 1, SEM_UNDO};

int GetSemID(key_t key)
{
    int sem_id = semget(sem_key, 1, IPC_CREAT | 0666);
    union semun sem_union = {0};
    sem_union.val = 0;
    if (semctl(sem_id, 0, SETVAL, sem_union) == -1) {
        std::cerr << "set sem failed" << std::endl;
        return -1;
    }
    return sem_id;
}

void FuseMounted()
{
    int sem_id = GetSemID(sem_key);
    semop(sem_id, &sem_unlock, 1);

    std::cout << "fuse_mounted" << std::endl;
}

int MountFile(const std::string &path, unsigned long offset, const std::string &mount_point)
{
    sem_key = getpid();
    int sem_id = GetSemID(sem_key);

    int fusefs_pid = fork();

    if (0 == fusefs_pid) {
        // child
        char *dir = realpath(path.c_str(), nullptr);

        char options[128];
        sprintf(options, "ro,offset=%lu", offset);

        const char *fuse_argv[] = {
            dir, "-o", options, dir, const_cast<char *>(mount_point.data()),
        };

        if (0 != fusefs_main(5, const_cast<char **>(fuse_argv), FuseMounted)) {
            std::cerr << "fusefs_main" << std::endl;
            return -1;
        }
    }

    // parent
    // wait mounted
    semop(sem_id, &sem_lock, 1);
    return 0;
}

} // namespace linglong::squashfs

src/uaploader.cpp
/*
 * Copyright (c) 2019-2021. Uniontech Software Ltd. All rights reserved.
 *
 * Author:     Heysion <heysoin@deepin.com>
 *
 * Maintainer: Heysion <heysion@deepin.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <filesystem>
#include <elf.h>
#include <sys/stat.h>
#include <cstring>
#include <unistd.h>
#include <sys/wait.h>
//#include <fuse.h>
//#include "squashfuse/ll.h"

using namespace std;

#if __BYTE_ORDER == __LITTLE_ENDIAN
#define ELFDATANATIVE ELFDATA2LSB
#elif __BYTE_ORDER == __BIG_ENDIAN
#define ELFDATANATIVE ELFDATA2MSB
#else
#error "Unknown machine endian"
#endif

#define bswap_16(value) \
((((value) & 0xff) << 8) | ((value) >> 8))

#define bswap_32(value) \
(((uint32_t)bswap_16((uint16_t)((value) & 0xffff)) << 16) | \
(uint32_t)bswap_16((uint16_t)((value) >> 16)))

#define bswap_64(value) \
(((uint64_t)bswap_32((uint32_t)((value) & 0xffffffff)) \
<< 32) | \
(uint64_t)bswap_32((uint32_t)((value) >> 32)))

template<typename P>
inline uint16_t file16_to_cpu(uint16_t val, const P &ehdr) {
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_16(val);
    return val;
}

template<typename P>
uint32_t file32_to_cpu(uint32_t val, const P &ehdr) {
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_32(val);
    return val;
}

template<typename P>
uint64_t file64_to_cpu(uint64_t val, const P &ehdr) {
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_64(val);
    return val;
}

auto read_elf64(FILE *fd, Elf64_Ehdr &ehdr) -> decltype(ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum)) {
    Elf64_Ehdr ehdr64;
    off_t ret = -1;

    fseeko(fd, 0, SEEK_SET);
    ret = fread(&ehdr64, 1, sizeof(ehdr64), fd);
    if (ret < 0 || (size_t) ret != sizeof(ehdr64)) {
        return -1;
    }

    ehdr.e_shoff = file64_to_cpu<Elf64_Ehdr>(ehdr64.e_shoff, ehdr);
    ehdr.e_shentsize = file16_to_cpu<Elf64_Ehdr>(ehdr64.e_shentsize, ehdr);
    ehdr.e_shnum = file16_to_cpu<Elf64_Ehdr>(ehdr64.e_shnum, ehdr);

    return (ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum));

}

auto get_elf_size(const std::string path) -> decltype(-1) {
    FILE *fd = nullptr;
    off_t size = -1;
    Elf64_Ehdr ehdr;

    fd = fopen(path.c_str(), "rb");
    if (fd == nullptr) {
        return -1;
    }
    auto ret = fread(ehdr.e_ident, 1, EI_NIDENT, fd);
    if (ret != EI_NIDENT) {
        return -1;
    }
    if ((ehdr.e_ident[EI_DATA] != ELFDATA2LSB) &&
        (ehdr.e_ident[EI_DATA] != ELFDATA2MSB)) {
        return -1;
    }
    if (ehdr.e_ident[EI_CLASS] == ELFCLASS32) {
        //size = read_elf32(fd);
    } else if (ehdr.e_ident[EI_CLASS] == ELFCLASS64) {
        size = read_elf64(fd, ehdr);
    } else {
        return -1;
    }
    fclose(fd);
    return size;
}

std::string run_exec(const std::string &cmd) {
    std::array<char, 1024> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) {
        return nullptr;
    }
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        result += buffer.data();
    }
    return result;
}

int main(int argc, char **argv) {
//    auto user_home = std::getenv("HOME");
//    auto user_name = std::getenv("USER");
    auto exec_path = std::filesystem::canonical("/proc/self/exe");

    if (!filesystem::exists(exec_path)) {
        cout << "load failed! " << exec_path << endl;
        exit(-1);
    }
    auto elf_fd = fopen(exec_path.c_str(), "rb");
    if (!elf_fd) {
        cout << "load failed! " << exec_path << endl;
        exit(-1);
    }

    // get elf size
    auto elf_size = get_elf_size(exec_path);
    struct stat elf_stat;
    if (stat(exec_path.c_str(), &elf_stat) != 0) {
        exit(-1);
    }
    if (elf_size >= elf_stat.st_size) {
        cout << "loader can not run for single!" << endl;
        exit(-1);
    }

    char temp_prefix[1024] = "/tmp/uap-XXXXXX";
    char *mount_point = mkdtemp(temp_prefix);

    // mount data.sqsfs
    std::string cmd = "squashfuse -o ro,offset=";
    cmd += std::to_string(elf_size);
    cmd += " ";
    cmd += exec_path;
    cmd += " ";
    cmd += mount_point;

    run_exec(cmd);
    std::string loader_path = mount_point;
    loader_path += "/loader";
    if (!std::filesystem::exists(loader_path)) {
        cout << "can not run!" << endl;
        exit(-1);
    }

    auto pid = fork();

    // run loader for child
    if (pid == 0) {
        if (chdir(mount_point) != 0) {
            cout << "change to work failed !" << endl;
            exit(-1);
        }

        int c = 1;
        char *load_argv[16] = {
                loader_path.data(),
                mount_point,
        };
        if (argc > 1 && argc < 16) {
            for (auto idx = 1; idx < argc; idx++) {
                c++;
                load_argv[c] = argv[idx];
            }
        }
        c++;
        load_argv[c] = nullptr;
        extern char **environ;
        // run loader shell
        auto ret = execve(loader_path.data(), load_argv, environ);
        if (ret != 0) {
            perror("run");
            cout << "run failed!" << endl;
        }
        exit(0);
    }
    if (pid < 0) {
        cout << "fork exec failed!" << endl;
        exit(-1);
    }
    waitpid(pid, 0, 0);
    // cleanup
    std::string umount_cmd = "fusermount -u ";
    umount_cmd += mount_point;
    run_exec(umount_cmd);
    rmdir(mount_point);
    return 0;
}

src/elf.cpp
/*
 * SPDX-FileCopyrightText: 2022 UnionTech Software Technology Co., Ltd.
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

#include <elf.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include <cstdio>
#include <cstdlib>
#include <iostream>


namespace linglong::elf {

#if __BYTE_ORDER == __LITTLE_ENDIAN
#define ELFDATANATIVE ELFDATA2LSB
#elif __BYTE_ORDER == __BIG_ENDIAN
#define ELFDATANATIVE ELFDATA2MSB
#else
#error "Unknown machine endian"
#endif

#define bswap_16(value) ((((value)&0xff) << 8) | ((value) >> 8))

#define bswap_32(value) \
    (((uint32_t)bswap_16((uint16_t)((value)&0xffff)) << 16) | (uint32_t)bswap_16((uint16_t)((value) >> 16)))

#define bswap_64(value) \
    (((uint64_t)bswap_32((uint32_t)((value)&0xffffffff)) << 32) | (uint64_t)bswap_32((uint32_t)((value) >> 32)))

template<typename P>
inline uint16_t file16_to_cpu(uint16_t val, const P &ehdr)
{
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_16(val);
    return val;
}

template<typename P>
uint32_t file32_to_cpu(uint32_t val, const P &ehdr)
{
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_32(val);
    return val;
}

template<typename P>
uint64_t file64_to_cpu(uint64_t val, const P &ehdr)
{
    if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
        val = bswap_64(val);
    return val;
}

auto read_elf64(FILE *fd, Elf64_Ehdr &ehdr) -> decltype(ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum))
{
    off_t ret = -1;

    fseeko(fd, 0, SEEK_SET);
    // read ELF header
    if (fread(&ehdr, 1, sizeof ehdr, fd) != sizeof(ehdr)) {
        return -1;
    }

    ehdr.e_shoff = file64_to_cpu(ehdr.e_shoff, ehdr);
    ehdr.e_shstrndx = file16_to_cpu(ehdr.e_shstrndx, ehdr);
    ehdr.e_shnum = file16_to_cpu(ehdr.e_shnum, ehdr);

    Elf64_Shdr shdr;

    // read section name string table
    // first, read its header
    fseek(fd, ehdr.e_shoff + ehdr.e_shstrndx * sizeof shdr, SEEK_SET);
    if (fread(&shdr, 1, sizeof shdr, fd) != sizeof(shdr)) {
        return -1;
    }

    // next, read the section, string data
    char *sh = (char *)malloc(shdr.sh_size);

    do {
        fseek(fd, shdr.sh_offset, SEEK_SET);
        if (fread(sh, 1, shdr.sh_size, fd) != shdr.sh_size) {
            break;
        }

        // read all section headers
        for (int idx = 0; idx < ehdr.e_shnum; idx++) {
            fseek(fd, ehdr.e_shoff + idx * sizeof(shdr), SEEK_SET);
            if (fread(&shdr, 1, sizeof(shdr), fd) != sizeof(shdr)) {
                break;
            }

            shdr.sh_name = file32_to_cpu(shdr.sh_name, ehdr);
            shdr.sh_offset = file32_to_cpu(shdr.sh_offset, ehdr);

            if (strcmp(sh + shdr.sh_name, ".bundle") == 0) {
                ret = shdr.sh_offset;
                break;
            }
        }
    } while (false);
    free(sh);
    return ret;
}

off_t get_bundle_offset(const std::string &path)
{
    FILE *fd;
    off_t size = -1;
    Elf64_Ehdr ehdr;

    fd = fopen(path.c_str(), "rb");
    if (fd == nullptr) {
        return -1;
    }

    do {
        auto ret = fread(ehdr.e_ident, 1, EI_NIDENT, fd);
        if (ret != EI_NIDENT) {
            break;
        }
        if ((ehdr.e_ident[EI_DATA] != ELFDATA2LSB) && (ehdr.e_ident[EI_DATA] != ELFDATA2MSB)) {
            break;
        }
        if (ehdr.e_ident[EI_CLASS] == ELFCLASS32) {
            // size = read_elf32(fd);
        } else if (ehdr.e_ident[EI_CLASS] == ELFCLASS64) {
            size = read_elf64(fd, ehdr);
        } else {
            break;
        }
    } while (false);

    fclose(fd);
    return size;
}

} // namespace linglong::elf
